CONTRIBUTING.md

```diff
@@ -145,32 +261,490 @@ for review, which allows you to sidestep this issue.
-6. When the backport pull request is merged and you have the necessary privileges you can also replace the label `9.needs: port to stable` with `8.has: port to stable` on the original pull request. This way maintainers can keep track of missing backports easier.
+- When changing the bootloader installation process, extra care must be taken. Grub installations cannot be rolled back, hence changes may break people’s installations forever. For any non-trivial change to the bootloader please file a PR asking for review, especially from \@edolstra.
 
-## Criteria for Backporting changes
+- Create a commit for each logical unit.
 
-Anything that does not cause user or downstream dependency regressions can be backported. This includes:
-- New Packages / Modules
-- Security / Patch updates
-- Version updates which include new functionality (but no breaking changes)
-- Services which require a client to be up-to-date regardless. (E.g. `spotify`, `steam`, or `discord`)
-- Security critical applications (E.g. `firefox`)
+- Check for unnecessary whitespace with `git diff --check` before committing.
+
+- If you have commits `pkg-name: oh, forgot to insert whitespace`: squash commits in this case. Use `git rebase -i`.
+
+- Format the commit messages in the following way:
+
+  ```
+  (pkg-name | nixos/<module>): (from -> to | init at version | refactor | etc)
+
+  (Motivation for change. Link to release notes. Additional information.)
+  ```
+
+  For consistency, there should not be a period at the end of the commit message's summary line (the first line of the commit message).
+
+  Examples:
+
+  * nginx: init at 2.0.1
+  * firefox: 54.0.1 -> 55.0
+
+    https://www.mozilla.org/en-US/firefox/55.0/releasenotes/
+  * nixos/hydra: add bazBaz option
+
+    Dual baz behavior is needed to do foo.
+  * nixos/nginx: refactor config generation
+
+    The old config generation system used impure shell scripts and could break in specific circumstances (see #1234).
+
+### Writing good commit messages
+
+In addition to writing properly formatted commit messages, it's important to include relevant information so other developers can later understand *why* a change was made. While this information usually can be found by digging code, mailing list/Discourse archives, pull request discussions or upstream changes, it may require a lot of work.
+
+Package version upgrades usually allow for simpler commit messages, including attribute name, old and new version, as well as a reference to the relevant release notes/changelog. Every once in a while a package upgrade requires more extensive changes, and that subsequently warrants a more verbose message.
+
+Pull requests should not be squash merged in order to keep complete commit messages and GPG signatures intact and must not be when the change doesn't make sense as a single commit.
```

Comment:
Squash merges merge the commit messages as well, not? And if the merger has a signing key configured, the new commit will be committed with that key. Of course the old ones get lost then. As far as I know there never was any agreement on this part.

Reply:
> And if the merger has a signing key configured, the new commit will be committed with that key

IIRC that's not true. Github would need access to that private key, wouldn't it? AFAIK it is only committed using the github signing key!

Reply:
Right!

Reply:
So effectively, all trust is removed in case of a squash-merge/rebase-merge because having a github signature does not require anything except a github account. At least for me, having github signing a commit is equal to not having a signature at all. YMMV of course.

Reply:
>  YMMV of course.

Exactly. Since commit signing has historically been a divisive issue I think we should not mention this requirement it in the contributing guidelines, at least not in this PR. A follow-up PR could cover that.

Reply:
Yeah, this sentence was previously also [in CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/25c37880ecac49691bfa4b5dcb3fc89ad80cb5b7/CONTRIBUTING.md#writing-good-commit-messages) so I don't think we need to discuss it here.